backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/bpartner/user/role/repository/UserRoleRepository.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * #%L
+ * de.metas.adempiere.adempiere.base
+ * %%
+ * Copyright (C) 2021 metas GmbH
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+package de.metas.bpartner.user.role.repository;
+
+import com.google.common.collect.ImmutableList;
+import de.metas.bpartner.user.role.UserAssignedRoleId;
+import de.metas.bpartner.user.role.UserRole;
+import de.metas.bpartner.user.role.UserRoleId;
+import de.metas.cache.CCache;
+import de.metas.user.UserId;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.ad.dao.IQueryBL;
+import org.compiere.model.I_C_User_Assigned_Role;
+import org.compiere.model.I_C_User_Role;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public class UserRoleRepository
+{
+	private final IQueryBL queryBL = Services.get(IQueryBL.class);
+
+	private final CCache<UserId, List<UserAssignedRoleId>> assignedUserRoleCache = CCache
+			.<UserId, List<UserAssignedRoleId>>builder()
+			.tableName(I_C_User_Assigned_Role.Table_Name)
+			.build();
+
+	private final CCache<UserRoleId, I_C_User_Role> userRoleCache = CCache
+			.<UserRoleId, I_C_User_Role>builder()
+			.tableName(I_C_User_Role.Table_Name)
+			.build();
+
+	@NonNull
+	public List<UserRole> getUserRoles(@NonNull final UserId userId)
+	{
+		final List<UserAssignedRoleId> assignedRoleIds = assignedUserRoleCache.getOrLoad(userId, this::getAssignedRoleIds);
+
+		return assignedRoleIds.stream()
```

Comment:
it's ok, assumingn that we don't have 1000s of roles, still, this might have been a good place to use the getAllOrLoad() method that i used in the issue description